feature/s3/transfermanager/api_op_PutObject.go

```diff
@@ -0,0 +1,1077 @@
+	o.Bucket = bucket
+	o.Key = aws.ToString(out.Key)
+	o.RequestCharged = RequestCharged(out.RequestCharged)
+	o.SSEKMSKeyID = aws.ToString(out.SSEKMSKeyId)
+	o.ServerSideEncryption = ServerSideEncryption(out.ServerSideEncryption)
+	o.VersionID = aws.ToString(out.VersionId)
+	o.ResultMetadata = out.ResultMetadata
+}
+
+// PutObject uploads an object to S3, intelligently buffering large
+// files into smaller chunks and sending them in parallel across multiple
+// goroutines. You can configure the chunk size and concurrency through the
+// Options parameters.
+//
+// Additional functional options can be provided to configure the individual
+// upload. These options are copies of the Uploader instance Upload is called from.
+// Modifying the options will not impact the original Uploader instance.
+//
+// It is safe to call this method concurrently across goroutines.
+func (c *Client) PutObject(ctx context.Context, input *PutObjectInput, opts ...func(*Options)) (*PutObjectOutput, error) {
+	i := uploader{in: input, options: c.options.Copy()}
+	for _, opt := range opts {
+		opt(&i.options)
+	}
+
+	return i.upload(ctx)
+}
+
+type uploader struct {
+	options Options
+
+	in *PutObjectInput
+
+	// PartPool allows for the re-usage of streaming payload part buffers between upload calls
+	partPool byteSlicePool
+}
+
+func (u *uploader) upload(ctx context.Context) (*PutObjectOutput, error) {
+	if err := u.init(); err != nil {
+		return nil, fmt.Errorf("unable to initialize upload: %w", err)
+	}
+	defer u.partPool.Close()
+
+	clientOptions := []func(o *s3.Options){
+		func(o *s3.Options) {
+			o.APIOptions = append(o.APIOptions,
+				middleware.AddSDKAgentKey(middleware.FeatureMetadata, userAgentKey),
+				addFeatureUserAgent,
+			)
+		}}
+
+	r, _, cleanUp, err := u.nextReader(ctx)
+
+	if err == io.EOF {
+		return u.singleUpload(ctx, r, cleanUp, clientOptions...)
+	} else if err != nil {
+		cleanUp()
+		return nil, err
+	}
+
+	mu := multiUploader{
+		uploader: u,
+	}
+	return mu.upload(ctx, r, cleanUp, clientOptions...)
+}
+
+func (u *uploader) init() error {
+	partSize := u.options.PartSizeBytes
+	if partSize < DefaultPartSizeBytes {
+		return fmt.Errorf("part size must be at least %d bytes", DefaultPartSizeBytes)
+	}
+	u.partPool = newByteSlicePool(partSize)
+	u.partPool.ModifyCapacity(u.options.Concurrency + 1)
+
+	return nil
+}
+
+func (u *uploader) singleUpload(ctx context.Context, r io.Reader, cleanUp func(), clientOptions ...func(*s3.Options)) (*PutObjectOutput, error) {
+	defer cleanUp()
+
+	params := u.in.mapSingleUploadInput(r, u.options.ChecksumAlgorithm)
+
+	out, err := u.options.S3.PutObject(ctx, params, clientOptions...)
+	if err != nil {
+		return nil, err
+	}
+
+	var output PutObjectOutput
+	output.mapFromPutObjectOutput(out, u.in.Bucket, u.in.Key)
+	return &output, nil
+}
+
+type httpClient interface {
+	Do(r *http.Request) (*http.Response, error)
+}
+
+type recordLocationClient struct {
+	httpClient
+	location string
```

Comment:
Is recording the location of the object on the client thread safe?

Reply:
I'm still not convinced that we need to support this feature...